Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp

```diff
@@ -104,19 +104,7 @@ void WebPermissionController::permissionChanged(WebCore::PermissionName permissi
         if (observer.descriptor().name != permissionName || observer.origin().topOrigin != topOrigin)
             return;
 
-        auto* context = observer.context();
-        if (!context)
-            continue;
-
-        auto source = WebCore::Permissions::sourceFromContext(*context);
-        // FIXME: Add support for workers.
-        ASSERT(source && *source == WebCore::PermissionQuerySource::Window);
-
-        auto* page = downcast<WebCore::Document>(context)->page();
-        if (!page)
-            continue;
-
-        query(WebCore::ClientOrigin { observer.origin() }, WebCore::PermissionDescriptor { permissionName }, page, *source, [observer = WeakPtr { observer }](auto newState) {
+        query(WebCore::ClientOrigin { observer.origin() }, WebCore::PermissionDescriptor { permissionName }, observer.page(), observer.source(), [observer = WeakPtr { observer }](auto newState) {
```

Comment:
If page() is null and source() is not ServiceWorker and SharedWorker, we might not need to query again.

Reply:
I believe that if page() is null and source() is not ServiceWorker and SharedWorker, then query() will fail (there is an assertion that checks this) since this is never supposed to happen.

Reply:
Assert does not mean return; it means crash

Reply:
It documents that something is impossible. Not sure we need a runtime check if something can't happen?

Reply:
The assert in WebPermissionController::query() holds when there is a null check for the page; since Rupin drops the check, there is no guarantee that the assert holds.